cupy/cuda/cub.pyx

```diff
@@ -73,19 +87,46 @@ def can_use_reduce_sum(x_dtype, dtype=None):
         # See _sum_auto_dtype in cupy/core/_routines_math.pyx for which dtypes
         # are promoted.
         support_dtype = [numpy.int64, numpy.uint64,
-                         numpy.float32, numpy.float64]
+                         numpy.float32, numpy.float64,
+                         numpy.complex64, numpy.complex128]
     elif dtype == x_dtype:
         support_dtype = [numpy.int8, numpy.uint8, numpy.int16, numpy.uint16,
                          numpy.int32, numpy.uint32, numpy.int64, numpy.uint64,
-                         numpy.float32, numpy.float64]
+                         numpy.float32, numpy.float64,
+                         numpy.complex64, numpy.complex128]
     else:
         return False
     if x_dtype not in support_dtype:
         return False
     return True
 
 
+# TODO(leofang): remove this hack when CUB supports complex numbers
 def reduce_min(core.ndarray x, out=None):
+    if x.dtype in (numpy.complex64, numpy.complex128):
+        # NumPy does the comparison of complex numbers in lexical order
+        # (numpy/numpy#2004): test real part first, then imaginary
+        out_re = _reduce_min(x.real, out if out is None else out.real)
+        idx = (x.real == out_re).nonzero()
+        if len(idx) == x.ndim:
```

Comment:
The condition looks always true.  Is the following intended?
```python
y = x[idx]
if len(y) == 1:
```

Reply:
I think your comment above is a different facet of your question:
> My concern on `reduce_(min|max)` is that the performance may depend on input data. Could you compare the performance for edge cases e.g. a large constant array `cupy.full((400, 32, 28, 28), 1+2j)`?

To answer them, first note that `len(idx) == x.ndim` is not always `True`. `len(idx)` is proportional to the number of elements whose real part is equal to `out_re`. So, in the worst case scenario as you suggested with a large constant array, `len(idx)==N*x.dim` (`N` is the total number of elements), and then we move on to compare the imaginary part of the entire array.

The `if` statement here is to handle the best case (with one hit). In a less ideal case (a few hits in the real part), the search space is greatly reduced in the next call of `reduce_(min|max)`.

So, yes, in the worst case we search through the `N`-element array twice, and in this case we solely rely on CUB's clever optimization. With the input array `x` being replaced by `x = cp.full((400, 32, 28, 28), 1+2j, dtype=dtype)`, we can see that while it takes a longer time on GPU (but CUB still outperforms the naive ufunc), the NumPy performance is even worse:

For `CUB_DISABLED=0`:
```shell
testing <class 'numpy.complex64'> + sum ...
cupy : 601.90087890625 ms
numpy : 11250.6796875 ms
testing <class 'numpy.complex64'> + max ...
cupy : 4847.3271484375 ms
numpy : 31283.07421875 ms
testing <class 'numpy.complex64'> + min ...
cupy : 4812.888671875 ms
numpy : 27644.96484375 ms
testing <class 'numpy.complex128'> + sum ...
cupy : 1170.836181640625 ms
numpy : 11166.1201171875 ms
testing <class 'numpy.complex128'> + max ...
cupy : 5515.779296875 ms
numpy : 28532.884765625 ms
testing <class 'numpy.complex128'> + min ...
cupy : 5516.4833984375 ms
numpy : 28526.55078125 ms
```
For `CUB_DISABLED=1`:
```shell
testing <class 'numpy.complex64'> + sum ...
cupy : 9120.3193359375 ms
numpy : 8410.7587890625 ms
testing <class 'numpy.complex64'> + max ...
cupy : 13439.9365234375 ms
numpy : 26741.98828125 ms
testing <class 'numpy.complex64'> + min ...
cupy : 13067.576171875 ms
numpy : 26748.564453125 ms
testing <class 'numpy.complex128'> + sum ...
cupy : 9805.685546875 ms
numpy : 10828.603515625 ms
testing <class 'numpy.complex128'> + max ...
cupy : 14115.16015625 ms
numpy : 27617.724609375 ms
testing <class 'numpy.complex128'> + min ...
cupy : 13888.92578125 ms
numpy : 27661.27734375 ms
```
I think this is the reason we wanna use CUB.